tutorials/llm/llama/domain-adaptive-pretraining/code/extend_tokenizer_utils.py

```diff
@@ -0,0 +1,286 @@
+    print("added dummy vocab: ", add_dummy_cnt)
+    print("new vocab_size: ", new_vocab_size)
+    print("padded vocab: ", 768)
+    print("total cnt (with padding vocab): ", total_cnt)
+    assert add_dummy_cnt >= 3, "there should be at least 3 extra tokens for finetuning"
+
+    record = []
+    N = len(m.pieces)
+    for i, sym in enumerate(new_tokens):
+        new_sym = m.SentencePiece()
+        new_sym.piece = sym
+        new_sym.score = 0.0  # default score for USER_DEFINED
+        new_sym.type = 4  # type value for USER_DEFINED
+        m.pieces.insert(N + i, new_sym)  # position after default control symbols ("<unk>", "<s>", "</s>")
+        record.append([sym, N + i])
+
+    N = len(m.pieces)
+    # extra_token_id = 3
+    # for i in range(total_cnt - N):
+    for i in range(add_dummy_cnt):
+        new_sym = m.SentencePiece()
+        new_sym.piece = f"<extra_id_{i}>"
+        new_sym.score = 0.0  # default score for USER_DEFINED
+        new_sym.type = 4  # type value for USER_DEFINED
+        m.pieces.insert(N + i, new_sym)  # position after default control symbols ("<unk>", "<s>", "</s>")
+        # extra_token_id += 1
+        record.append([new_sym.piece, N + i])
+
+    with open(new_vocab_path, "w", encoding="utf8") as fp:
+        json.dump(record, fp)
+
+    with open(new_model_path, 'wb') as f:
+        f.write(m.SerializeToString())
+
+    print("token_cnt with customized tokenizer: ")
+    sp = spm.SentencePieceProcessor()
+    sp.load(new_model_path)
+    get_token_cnt_spm(data_root, sp, batchsize=1000, keys=["text"])
+
+    old_ebd_paths = []
+    for f in glob.glob(old_ebd_path + "/*.pt"):
+        old_ebd_paths.append(f)
+
+    def myFunc(s):
+        return int(s.split("embedding_")[-1].split(".")[0])
+
+    old_ebd_paths.sort(key=myFunc)
+    word_embeddings = []
+    output_layers = []
+    for f in old_ebd_paths:
+        temp = torch.load(f)
+        word_embeddings.append(temp['word_embeddings'])
+        output_layers.append(temp['output_layer'])
+    word_embedding = torch.cat(word_embeddings, dim=1)
+    output_layer = torch.cat(output_layers, dim=0)
+    print("word_embedding shape: ", word_embedding.shape)
+    print("output_layer shape: ", output_layer.shape)
+
+    N_ori_emb, N = word_embedding.shape
+    # add_cnt = total_cnt - N_ori_emb
+    add_weight = torch.zeros(total_add_cnt, N)
+    word_embedding = torch.cat((word_embedding[:ori_vocab_size], add_weight, word_embedding[ori_vocab_size:]), 0)
+
+    _, M = output_layer.shape
+    add_out = torch.zeros(total_add_cnt, M)
+    output_layer = torch.cat((output_layer[:ori_vocab_size], add_out, output_layer[ori_vocab_size:]), 0)
+
+    sp = spm.SentencePieceProcessor()
+    sp.load(original_tokenizer_path)
+
+    for r in record:
+        token = r[0]
+        idx = r[1]
+        ids = sp.encode_as_ids(token)
+        word_embedding[idx] = torch.mean(word_embedding[ids], dim=0)
+        output_layer[idx] = torch.mean(output_layer[ids], dim=0)
```

Comment:
"The weights in the output layer corresponding to the new token are initialized to zero." according to the notebook?

Reply:
Thanks for pointing that out. I have updated the notebook to reflect the mean initialization.